src/apps/polymer/IncentivizedPolymerEscrow.sol

```diff
@@ -0,0 +1,160 @@
+    ) external payable override {
+        revert NotImplemented();
+    }
+
+    // packet.srcPortAddr is the IncentivizedPolymerEscrow address on the source chain.
+    // packet.destPortAddr is the address of this contract.
+    // channelId: the universal channel id on the source chain, which can be used to identify the source chain.
+    function onRecvUniversalPacket(bytes32 channelId, UniversalPacket calldata packet)
+        external
+        onlyIbcMw
+        returns (AckPacket memory)
+    {
+        uint256 gasLimit = gasleft();
+        bytes32 feeRecipitent = bytes32(uint256(uint160(tx.origin)));
+
+        bytes memory sourceImplementationIdentifier = abi.encodePacked(packet.srcPortAddr);
+
+        bytes memory receiveAck = _handleMessage(
+            bytes32(abi.encodePacked(packet.srcPortAddr)),
+            sourceImplementationIdentifier,
+            packet.appData,
+            feeRecipitent,
+            gasLimit
+        );
+
+        // Send ack:
+        return AckPacket({success: true, data: receiveAck});
+    }
+
+    // The escrow manages acks, so any message can be directly provided to _handleAck.
+    function onUniversalAcknowledgement(UniversalPacket calldata packet, AckPacket calldata ack) external onlyIbcMw {
+        uint256 gasLimit = gasleft();
+        bytes32 feeRecipitent = bytes32(uint256(uint160(tx.origin)));
+
+        bytes calldata rawMessage = ack.data;
+        // TODO: get dest chain ID?
+        bytes32 chainIdentifier;
+        bytes memory destinationImplementationIdentifier = abi.encodePacked(packet.destPortAddr);
+
+        isVerifiedMessageHash[keccak256(rawMessage)] = VerifiedMessageHashContext({
+            chainIdentifier: chainIdentifier,
+            implementationIdentifier: destinationImplementationIdentifier
+        });
+        _handleAck(chainIdentifier, destinationImplementationIdentifier, rawMessage, feeRecipitent, gasLimit);
+    }
+
+    // For timeouts, we need to construct the message.
+    function onTimeoutUniversalPacket(UniversalPacket calldata packet) external onlyIbcMw {
+        uint256 gasLimit = gasleft();
+        bytes32 feeRecipitent = bytes32(uint256(uint160(tx.origin)));
+
+        bytes calldata rawMessage = packet.appData;
+        // TODO: get dest chain ID?
+        bytes32 chainIdentifier;
+
+        _handleTimeout(chainIdentifier, rawMessage, feeRecipitent, gasLimit);
+    }
+
+    // * Send to messaging_protocol
+    function _sendPacket(
+        bytes32 destinationChainIdentifier,
+        bytes memory, /* destinationImplementation */
+        bytes memory message
+    ) internal override returns (uint128 costOfsendPacketInNativeToken) {
+        // TODO: get Polymer universal channelId from Polymer registry. Each channelID is unique for a pair of chains.
+        bytes32 channelId = _getChannelId(destinationChainIdentifier);
+        // TODO: get IncentivizedPolymerEscrow address deployed on the destination chain.
+        address destEscrowAddr;
```

Comment:
You can use destinationImplementation here. Say:
```
destEscrowAddr = address(uint160(uint256(bytes32(destinationImplementation)));
```
It is assumed that the destinationImplementation is provided as intended for the AMB.

However, you might consider making destEscrowAddr be bytes32, bytes, or something larger to account for chains with larger than 20 bytes addresses.

Reply:
Good point! I updated our `vibc-core-smart-contracts` repo so packet source and destination addresses are all `bytes32` types